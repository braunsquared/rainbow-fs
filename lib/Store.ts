import aguid from 'aguid';
import { BucketTree } from './BucketTree';
import { Bucket } from './Bucket';

import { IDb, DatabaseFactory } from './model/Database';
import { ItemFactory, IItem, ItemMetaData } from './model/Sitecore';
import { memoryDbFactory } from './impl/db/MemoryDb';
import { defaultItemFactory } from './impl/sitecore/Item';
import { Path } from './impl/sitecore/Path';
import { Minimatch, IMinimatch } from 'minimatch';

export type IDSource = (objOrItem: any) => string;

export interface IStoreConfig {
  dbFactory: DatabaseFactory;
  itemFactory: ItemFactory;
  idSource: IDSource;
}

export interface IStore {
  readonly config: IStoreConfig;

  addBucket(bucket: Bucket): void;

  getDatabase(name: string): IDb | undefined;
  getOrCreateDatabase(name: string): IDb;

  generateId(src: string): string;
  createItemFromObject(obj: any, meta?: ItemMetaData): IItem;

  glob(pattern: string, db?: string | string[]): IItem[];

  readAll(): Promise<any>;
}

export const defaultConfig: IStoreConfig = {
  dbFactory: memoryDbFactory,
  itemFactory: defaultItemFactory,
  idSource: item => item.Path instanceof Path ? item.Path.Path : item.Path,
}

/**
 * The Store is the root container that ties everything together.  It manages
 * the database, buckets and factories for loading/creating/mutating items in
 * the data store.
 */
export class Store implements IStore {
  private _trees: Map<string, BucketTree>;
  private _dbs: Map<string, IDb>;
  private _config: IStoreConfig;

  constructor(config: IStoreConfig = defaultConfig) {
    this._config = {...defaultConfig, ...config};

    this._trees = new Map();
    this._dbs = new Map();
  }

  /**
   * @returns The current `Store` configuration
   */
  get config() {
    return this._config;
  }

  /**
   * @returns A map of all the root `BucketTree`. One per `IDb` instance keyed on the database name.
   */
  get trees() {
    return this._trees;
  }

  /**
   * Get an existing `IDb` instance
   * 
   * @param name - The database name
   * @returns The `IDb` instance or undefined
   */
  getDatabase(name: string): IDb | undefined {
    return this._dbs.get(name);
  }

  /**
   * Get an existing `IDb` instance or create one using the `dbFactory` if it does not exist.
   * 
   * @param name The database name
   * @returns The `IDb` instance
   */
  getOrCreateDatabase(name: string): IDb {
    let db = this._dbs.get(name);
    if(!db) {
      db = this._config.dbFactory(name);
      this._dbs.set(name, db);
    }
    return db;
  }

  /**
   * Register a `Bucket` with it's respective database `BucketTree`.
   * 
   * @param bucket - The `Bucket` to add
   */
  addBucket(bucket: Bucket): void {
    let tree = this._trees.get(bucket.DB);
    if (!tree) {
      tree = new BucketTree();
      this._trees.set(bucket.DB, tree);
    }
    tree.addBucket(bucket);
  }

  /**
   * Create a new `IItem` using the `itemFactory` and register it with it's respective `IDb`
   * 
   * @param obj - The deserialized item data as a plain JSON Object.
   * @param meta - Any meta data to associate with the created `IItem`
   */
  createItemFromObject(obj: any, meta?: ItemMetaData): IItem {
    const item = this.config.itemFactory(this, obj, meta);
    const db = this.getOrCreateDatabase(item.DB);
    return db.addItem(item);
  }

  /**
   * Generate a new deterministic GUID using `src` as the seed value.
   * 
   * @param src - The source string for the generated GUID
   * @returns A predicatable GUID
   */
  generateId(src: string): string {
    return aguid(src);
  }

  private _globMatch(matcher: IMinimatch, dbName: string): IItem[] {
    const db = this._dbs.get(dbName);
    if (!db) {
      throw new Error(`DB not found: ${dbName}`);
    }

    const items: IItem[] = [];
    db.items.forEach(i => {
      if (matcher.match(i.Path.Path)) {
        items.push(i);
      }
    });

    return items;
  }

  /**
   * Find items in the `Store` which match the Glob pattern.  If no dbs are passed, a match
   * against all the databases is performed.
   * 
   * @param pattern - A standard Glob pattern
   * @param dbs - A database name or array of database names to scope the search to
   * @returns An array of `IItem`s containing items that matched the Glob pattern
   */
  glob(pattern: string, dbs?: string | string[]): IItem[] {
    const results: IItem[] = [];
    const matcher = new Minimatch(pattern, { nocase: true });

    let dbsToTest;

    if (!dbs) {
      dbsToTest = Array.from(this._dbs.keys());
    } else {
      dbsToTest = Array.isArray(dbs) ? dbs : [dbs];
    }

    for(let i = 0; i < dbsToTest.length; ++i) {
      results.splice(results.length, 0, ...this._globMatch(matcher, dbsToTest[i]));
    }

    return results;
  }

  /**
   * Perform a `read` call on all registered `Bucket`s
   */
  async readAll() {
    const promises: Promise<any>[] = [];
    this._trees.forEach(tree => {
      tree.walk(bucket => {
        promises.push(bucket.read(this));
      });
    });
    await Promise.all(promises);
  }
}