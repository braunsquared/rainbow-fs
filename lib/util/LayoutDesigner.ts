import { GUID } from '../model/Sitecore';

export interface IDeviceDefinition {
  ID: GUID;
  Layout: string;
  Placeholders: IPlaceholderDefinition[];
  Renderings: IRenderingDefinition[];
}

export interface ILayoutDefinition {
  Devices: IDeviceDefinition[];
}

export interface IRenderingDefinition {
  UID: GUID;
  ItemID: GUID;
  Cachable: string;
  Conditions: string;
  Datasource: string;
  MultiVariateTest: string;
  PersonalizationTest: string;
  Parameters: string;
  Placeholder: string;
  VaryByData: string;
  VaryByDevice: string;
  VaryByLogin: string;
  VaryByParameters: string;
  VaryByQueryString: string;
  VaryByUser: string;
  ClearOnIndexUpdate: string;
}

export interface IPlaceholderDefinition {
  UID: GUID;
  Key: string;
  MetaDataItemID: GUID;
}

export class LayoutDesigner {

}
