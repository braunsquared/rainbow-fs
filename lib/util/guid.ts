import { GUID } from '../model/Sitecore';

export function toGUID(val: string): GUID {
  const match = /^\{?([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})\}?$/.exec(val);
  if (match) {
    return match[1].toLowerCase();
  }
}

export function compareGUIDs(a: GUID, b: GUID): boolean {
  if (!a || !b) {
    return false
  } else {
    return a.toLowerCase() === b.toLowerCase();
  }
}