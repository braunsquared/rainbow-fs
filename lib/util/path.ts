
export const sep = '/';

export function normalize(path: string): string {
  const norm = path.replace(/\/{2,}/g, '/').replace(/\/$/, '');
  return norm.startsWith('/') ? norm : `/${norm}`;
}

export function join(...elements: (string | undefined)[] ): string {
  return normalize(elements.filter(i => !!i).join(sep));
}

export function split(path: string): string[] {
  const tokens = path.split(sep);
  return tokens.length > 0 && tokens[0] === '' ? tokens.slice(1) : tokens;
}

export interface ParsedPath {
  name: string,
  base: string,
  path: string,
  tokens: string[],
};

export function parse(path: string): ParsedPath {
  const norm = normalize(path);
  const tokens = split(norm);
  return {
    name: tokens[tokens.length - 1],
    base: join(...tokens.slice(0, tokens.length -1)),
    path: norm,
    tokens,
  };
}