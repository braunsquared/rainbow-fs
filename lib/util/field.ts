import { GUID, IField } from '../model/Sitecore';
import { toGUID } from './guid';

export function treelistValue(field: IField | undefined, defaultValue: GUID[] = []): GUID[] {
  if (!field || !field.Value) {
    return defaultValue;
  }

  return field.Value.split(/\n|\r\n/).map(id => toGUID(id)).filter(i => i);
}

export function booleanValue(field: IField | undefined, defaultValue: boolean = false): boolean {
  if (!field || !field.Value) {
    return defaultValue;
  }

  return field.Value === '1';
}