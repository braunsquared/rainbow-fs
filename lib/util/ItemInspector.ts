import { IStore } from '../Store';
import { IItem, GUID } from '../model/Sitecore';
import { treelistValue, booleanValue } from './field';
import { compareGUIDs } from './guid';

export interface FieldDefinition {
  ID: GUID;
  Name: string;
  Type: string;
  Shared: boolean;
  Unversioned: boolean;
  Item?: IItem;
}

export interface FieldDefinitionResults {
  errors: Error[];
  fields: FieldDefinition[];
}

class ItemInspector {
  static SystemTemplateGUIDs = [
    '1930bbeb-7805-471a-a3be-4858ac7cf696', // Standard Template
    'a87a00b1-e6db-45ab-8b54-636fec3b5523', // Folder
    'b36ba9fd-0dc0-49c8-bea2-e55d70e6af28', // JSS Route
  ];

  private _store: IStore;
  private _item: IItem;

  constructor(store: IStore, item: IItem) {
    this._store = store;
    this._item = item;
  }

  get Item(): IItem { return this._item; }

  getTemplateItem(): IItem | undefined {
    const db = this._store.getDatabase(this._item.DB);
    if (!db) {
      throw new Error(`Item database not found: ${this._item.DB}`);
    }

    return db.getItem(this._item.Template);
  }

  getStandardValues(): IItem | undefined {
    let standardVals = this._store.glob(`${this._item.Path.Path}/__Standard Values`, this._item.DB);
    standardVals = standardVals.filter(item => item.Parent === this._item.ID);
    return standardVals[0];
  }

  inspectTemplate(): ItemInspector {
    const tmpl = this.getTemplateItem();
    if (!tmpl) {
      throw new Error(`Template item not found: ${this._item.Template}`);
    }

    return new ItemInspector(this._store, tmpl);
  }

  children(): IItem[] {
    const children = this._store.glob(`${this._item.Path.Path}/*`, this._item.DB);
    return children.filter(i => i.Parent === this._item.ID);
  }

  fieldDefinitions(): FieldDefinitionResults {
    const result: FieldDefinitionResults = {
      errors: [],
      fields: [],
    };

    const db = this._store.getDatabase(this._item.DB)!;

    const generateFieldDefinitions = (inspector: ItemInspector) => {
      const sections = inspector.children();
      sections.forEach(sectionItem => {
        const itemInspect = new ItemInspector(this._store, sectionItem);
        const fields = itemInspect.children();
        fields.forEach(fieldItem => {
          const typeField = fieldItem.field('Type');
          const sharedField = fieldItem.field('Shared');
          const unversionField = fieldItem.field('Unversioned');

          result.fields.push({
            ID: fieldItem.ID,
            Name: fieldItem.Path.Name,
            Type: typeField && typeField.Value ? typeField.Value : '',
            Shared: booleanValue(sharedField),
            Unversioned: booleanValue(unversionField),
            Item: fieldItem,
          });
        });
      });
    }

    // Get the Base Templates
    const baseField = this._item.field('__Base template');
    if(baseField && baseField.Value) {
      const templateIds = treelistValue(baseField);
      const baseInspectors = templateIds.map(id => {
        const item = db.getItem(id);
        if (!item) {
          // Only push errors for items that aren't the standard template id
          if (!id || (id && ItemInspector.SystemTemplateGUIDs.indexOf(id.toLowerCase()) === -1)) {
            result.errors.push(new Error(`Unable to find base template: ${id}`));
          }
          return;
        }

        return new ItemInspector(this._store, item);
      });
      baseInspectors.forEach(inspect => inspect && generateFieldDefinitions(inspect));
    }

    // Get the field definitions for the template item
    generateFieldDefinitions(this);

    return result;
  }
}

export { ItemInspector };
