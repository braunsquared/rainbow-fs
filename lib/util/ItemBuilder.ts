import { IStore } from '../Store';
import { IItem, GUID } from '../model/Sitecore';
import { ItemInspector, FieldDefinition } from './ItemInspector';
import { compareGUIDs } from './guid';
import { Path } from '../impl/sitecore/Path';
import aguid from 'aguid';
import * as luxon from 'luxon';

export interface Blob {
  Value: string;
  BlobID: string;
}

export class ItemBuilder {
  private _store: IStore;
  private _templateId: GUID;

  private _fields: Map<FieldDefinition, any>;
  private _blobs: Map<FieldDefinition, Blob>;
  private _knownDefs: FieldDefinition[];

  constructor(store: IStore, templateItem: IItem | string) {
    this._store = store;
    this._fields = new Map();
    this._blobs = new Map();

    if (typeof templateItem === 'string') {
      this._templateId = templateItem;
      this._knownDefs = [];
    } else {
      this._templateId = templateItem.ID;
      const inspector = new ItemInspector(store, templateItem);
      this._knownDefs = inspector.fieldDefinitions().fields;
    }
  }

  setFieldByName(name: string, value: any): this {
    const def = this._knownDefs.find(def => def.Name === name);
    if (!def) {
      throw new Error(`Unknown field: ${name}`);
    }
    return this.setField(def, value);
  }

  setFieldByGUID(fieldId: GUID, value: any): this {
    const def = this._knownDefs.find(def => compareGUIDs(def.ID, fieldId));
    if (!def) {
      throw new Error(`Unknown field: ${fieldId}`);
    }
    return this.setField(def, value);
  }

  // tslint:disable-next-line:max-line-length
  setFieldWithDetails(fieldId: GUID, name: string, type: string, shared: boolean, unversioned: boolean, value: any): this {
    return this.setField(
      {
        ID: fieldId,
        Name: name,
        Type: type,
        Shared: shared,
        Unversioned: unversioned,
      },
      value,
    );
  }

  setField(def: FieldDefinition, value: any): this {
    for (const key of this._fields.keys()) {
      if (compareGUIDs(key.ID, def.ID) && key !== def) {
        throw new Error(`Duplicate field key: ${key.ID}`);
      }
    }

    this._fields.set(def, value);
    return this;
  }

  setBlobByName(name: string, base64: string): this {
    const def = this._knownDefs.find(def => def.Name === name);
    if (!def) {
      throw new Error(`Unknown field: ${name}`);
    }
    return this.setBlob(def, base64);
  }

  setBlobByGUID(fieldId: GUID, base64: string): this {
    const def = this._knownDefs.find(def => compareGUIDs(def.ID, fieldId));
    if (!def) {
      throw new Error(`Unknown field: ${fieldId}`);
    }
    return this.setBlob(def, base64);
  }

  // tslint:disable-next-line:max-line-length
  setBlobWithDetails(fieldId: GUID, name: string, type: string, shared: boolean, unversioned: boolean, base64: string): this {
    return this.setBlob(
      {
        ID: fieldId,
        Name: name,
        Type: type,
        Shared: shared,
        Unversioned: unversioned,
      },
      base64,
    );
  }

  setBlob(def: FieldDefinition, base64: string): this {
    const blobId = aguid(base64);
    this._blobs.set(def, { Value: base64, BlobID: blobId });

    return this;
  }

  private _serializeField = (def: FieldDefinition): any => {
    return {
      ID: def.ID,
      Hint: def.Name,
      Value: this._fields.get(def),
    };
  }

  private _serializeBlob = (def: FieldDefinition): any => {
    const value = this._blobs.get(def);

    if (!value) return;

    return {
      ID: def.ID,
      Hint: def.Name,
      BlobID: value.BlobID,
      Value: value.Value,
    };
  }

  private _serializeFields = (predicate: (def: FieldDefinition) => boolean): any[] => {
    const fieldDefs = Array.from(this._fields.keys()).filter(predicate).map(this._serializeField);
    const blobDefs = Array.from(this._blobs.keys()).filter(predicate).map(this._serializeBlob);
    return [...fieldDefs, ...blobDefs];
  }

  buildItem(dbName: string, scPath: string, language?: string): IItem {
    // We build the item js and then let the store item factory do the magic
    const path = new Path(scPath);

    const itemJs: any = {
      Name: path.Name,
      Template: this._templateId,
      Path: path.Path,
      DB: dbName,
    };

    const sharedFields = this._serializeFields(def => def.Shared);
    if (sharedFields.length) {
      itemJs.SharedFields = sharedFields;
    }

    if (language) {
      const langJs: any = { Language: language };
      const unversionedFields = this._serializeFields(def => def.Unversioned);
      if (unversionedFields.length) {
        langJs.Fields = unversionedFields;
      }

      const versionJs: any = { Version: 1 };
      const versionedFields = this._serializeFields(def => !def.Shared && !def.Unversioned);
      if (versionedFields.length) {
        versionJs.Fields = versionedFields;
      }
      langJs.Versions = [versionJs];
      itemJs.Languages = [langJs];
    }

    return this._store.createItemFromObject(itemJs, {
      isNew: true,
      isDirty: true,
      isVirtual: false,
    });
  }

  static formatDate(date: Date): string {
    return luxon.DateTime.fromJSDate(date).toFormat('yyyyMMdd\'T\'HHmmss\'Z\'');
  }
}
