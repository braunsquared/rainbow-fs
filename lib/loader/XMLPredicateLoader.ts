import parser from 'fast-xml-parser';
import fs from 'fs';
import path from 'path';
import util from 'util';

import { IStore } from '../Store';
import { Bucket } from '../Bucket';

const readFile = util.promisify<string, Buffer>(fs.readFile);

function __asArray(val: any) {
  return Array.isArray(val) ? val : [val];
}

class XMLPredicateLoader {
  private _variables: Map<string, string>;

  constructor() {
    this._variables = new Map();
  }

  setVariable(key: string, value: string) {
    this._variables.set(key, value);
  }

  replaceVariables(input: string): string {
    return input.replace(/\$\(([^)]*)\)/g, (_, key) => {
      return this._variables.get(key) || ''
    });
  }

  async load(xmlPath: string, store: IStore) {
    const data = await readFile(xmlPath);
    const xmlData = data.toString('utf8');
    const json = parser.parse(xmlData, {
      attributeNamePrefix: '',
      ignoreAttributes: false,
      attrNodeName: 'attr',
      allowBooleanAttributes: true,
      parseAttributeValue: true,
    });

    // Find the unicorn config
    const unicorn = json.configuration.sitecore.unicorn;

    // Iterate over the configurations
    const configs = __asArray(unicorn.configurations.configuration);

    configs.forEach(config => {
      // Determine the fspath
      const physicalRoot = this.replaceVariables(config.targetDataStore.attr.physicalRootPath);

      // Add the includes
      const includes = __asArray(config.predicate.include);
      includes.forEach(include => {
        const fsPath = path.join(physicalRoot, include.attr.name);
        const db = include.attr.database;
        const scPath = include.attr.path;
        store.addBucket(new Bucket(fsPath, scPath, db));
      });
    });
  }
}

export { XMLPredicateLoader };