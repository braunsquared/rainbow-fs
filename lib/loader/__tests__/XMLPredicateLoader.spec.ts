import path from 'path';
import mockfs from 'mock-fs';
import { Store, IStore } from '../../Store';
import { XMLPredicateLoader } from '../XMLPredicateLoader';
import { loadMockFs } from '../../../test/util';

describe('XMLPredicateLoader', () => {
  let store: IStore;
  let loader: XMLPredicateLoader;
  let sampleData: any;

  beforeAll(() => {
    sampleData = loadMockFs();
  });

  beforeEach(() => {
    mockfs(sampleData);
    store = new Store();
    loader = new XMLPredicateLoader();
  });

  afterEach(() => {
    mockfs.restore();
  })

  it('should load predicates from an xml config file', () => {
    loader.setVariable('dataFolder', '');
    return loader.load('/xml/SampleData.Serialization.config', store).then(() => {
      mockfs.restore();
      expect((<Store>store).trees).toMatchSnapshot();
    });
  });
});