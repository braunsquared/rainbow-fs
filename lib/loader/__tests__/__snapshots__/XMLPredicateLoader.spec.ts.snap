// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XMLPredicateLoader should load predicates from an xml config file 1`] = `
Map {
  "master" => BucketTree {
    "root": BucketEntry {
      "bucket": null,
      "children": Array [
        BucketEntry {
          "bucket": Bucket {
            "DB": "master",
            "Filepath": "data\\\\Renderings",
            "Path": Path {
              "Folder": "/sitecore/layout/Renderings/Helix/Project",
              "Name": "TestProject",
              "Path": "/sitecore/layout/Renderings/Helix/Project/TestProject",
              "Tokens": Array [
                "sitecore",
                "layout",
                "Renderings",
                "Helix",
                "Project",
                "TestProject",
              ],
            },
          },
          "children": Array [],
          "parent": [Circular],
          "path": "/sitecore/layout/Renderings/Helix/Project/TestProject",
        },
        BucketEntry {
          "bucket": Bucket {
            "DB": "master",
            "Filepath": "data\\\\Templates",
            "Path": Path {
              "Folder": "/sitecore/templates/Helix/Project",
              "Name": "TestProject",
              "Path": "/sitecore/templates/Helix/Project/TestProject",
              "Tokens": Array [
                "sitecore",
                "templates",
                "Helix",
                "Project",
                "TestProject",
              ],
            },
          },
          "children": Array [],
          "parent": [Circular],
          "path": "/sitecore/templates/Helix/Project/TestProject",
        },
      ],
      "parent": null,
      "path": "",
    },
  },
}
`;
